### PR TITLE
Improve safe chunk validation

### PR DESCRIPTION
**Pull request by josephnowak**
@max-sixty Here is the PR with all the features requested https://github.com/pydata/xarray/pull/9513#discussion_r1767677435, also I added more tests and improved the logic of the validation algorithm, I think now it is much more simple and covers certain cases that were not validated before, for example, the "r+" mode only allows full chunk writes even on the last chunk.
- [ ] Closes #xxxx
- [test_zarr_safe_chunk_region ] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: xarray/backends/zarr.py [hunk 1]
```diff
@@ -189,20 +193,48 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name, safe_chunks):
...
+
+            # The first border size is the amount of data that needs to be updated on the
+            # first chunk taking into account the region slice.
+            first_border_size = zchunk
+            if interval.start:
+                first_border_size = zchunk - interval.start % zchunk
+
+            if not allow_partial_chunks and first_border_size < zchunk:
+                # If the border is smaller than zchunk, then it is a partial chunk write
+                raise ValueError(first_border_size)
```
Comment (max-sixty) [inline xarray/backends/zarr.py:228]: Maybe a few words explaining the error?

File: xarray/backends/zarr.py [hunk 2]
```diff
@@ -243,15 +275,24 @@ def _get_zarr_dims_and_attrs(zarr_obj, dimension_key, try_nczarr):
 
 
 def extract_zarr_variable_encoding(
-    variable, raise_on_invalid=False, name=None, safe_chunks=True
+    variable,
+    raise_on_invalid=False,
+    name=None,
+    safe_chunks=True,
```
Comment (max-sixty) [inline xarray/backends/zarr.py:281]:
```suggestion
    *,
    safe_chunks=True,
```

File: xarray/backends/zarr.py [hunk 3]
```diff
@@ -243,15 +275,24 @@ def _get_zarr_dims_and_attrs(zarr_obj, dimension_key, try_nczarr):
...
+    region=None,
+    mode=None,
 ):
     """
     Extract zarr encoding dictionary from xarray Variable
 
     Parameters
     ----------
     variable : Variable
+    region: tuple[slice], optional
```
Comment (max-sixty) [inline xarray/backends/zarr.py:291]: Could we order the same as the order of params?

File: xarray/backends/zarr.py [hunk 4]
```diff
@@ -189,20 +193,48 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name, safe_chunks):
...
-                            + " Consider either rechunking using `chunk()`, deleting "
-                            "or modifying `encoding['chunks']`, or specify `safe_chunks=False`."
-                        )
+        # If it is possible to write on partial chunks then it is not necessary to check
+        # the last one contained on the region
+        allow_partial_chunks = True
+        end = -1
+        if mode == "r+":
+            # This mode forces to write only on full chunks, even on the last one
+            allow_partial_chunks = False
```
Comment (max-sixty) [inline xarray/backends/zarr.py:202]: Nit: could we hoist the `allow_partial_chunks` out of the loop and just have `allow_partial_chunks = mode != "r+"`

**Review by max-sixty — APPROVED** [review]
This looks really good @josephnowak ! The tests look great. 

I left a couple of small comments and will have another read through of the code later, but overall 👍  . Thank you.

File: xarray/backends/zarr.py (hunk 1 above)
Comment (josephnowak) [reply xarray/backends/zarr.py:228]: Sorry, it was supposed to report the same error than in the loop

Comment (josephnowak) [reply xarray/backends/zarr.py]: I already applied all the modifications on the last commit

**Comment by max-sixty** [comment]
Couple of typing errors, then we can merge!

**Comment by josephnowak** [comment]
Hi @max-sixty, sorry for being insistent with the review but I would like to avoid any misunderstood with the validation logic and affect the usability of Xarray. The following case is going to raise an error using the mode "r+":
```python
    arr = xr.DataArray(
        list(range(10)), dims=["a"], coords={"a": list(range(10))}, name="foo"
    ).chunk(a=3)
    arr.to_zarr(store, mode="w")
    with pytest.raises(ValueError):
        # It is not possible to write on a chunk that is not full, even when it is the last one which can be of a smaller size
        arr.isel(a=slice(9, 10)).to_zarr(store, region="auto", mode="r+")
```
Is it the expected behavior? if it is, it probably would be good to make "a" the default mode, because it allows partial writes on the first and last chunk inside the region which I think is more convenient in more scenarios

**Comment by max-sixty** [comment]
> Hi @max-sixty, sorry for being insistent with the review but I would like to avoid any misunderstood with the validation logic and affect the usability of Xarray. 

(tbc, thank you very much for being insistent, this is a complicated and important problem which we haven't spent enough time on, so need exactly this sort of help!)

> The following case is going to raise an error using the mode "r+":

OK, that's a very good point — that writing to the full final chunk will raise an error with the proposed code. The rule should be that partial chunk writes are not allowed, but that is a full chunk write, so should be allowed. Otherwise people can't write to the full array!

Could we adjust the logic so writing to the final chunk is allowed?

> if it is, it probably would be good to make "a" the default mode, because it allows partial writes on the first and last chunk inside the region which I think is more convenient in more scenarios

I weigh the cost of unsafe writes very highly, so my initial impulse is that writing to partial chunks should fail by default.

Possibly making this strict could break some existing code; possibly we should explain ourselves better. For users who want to write to a single partial chunk from a single process, it would be a fairly small change for people to pass `mode="a"`, or allow unsafe writes. 

But open to feedback from others; CC @pydata/xarray. One compromise would be to have a new mode, like `rd+`, for distributed writes, which would be strict like this. (though we should only do that if people feel strongly, otherwise we're at risk of config proliferation)

File: xarray/backends/zarr.py [hunk 5]
```diff
@@ -189,20 +193,54 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name, safe_chunks):
...
+            f"Writing this array in parallel with dask could lead to corrupted data."
+            f"Consider either rechunking using `chunk()`, deleting "
+            f"or modifying `encoding['chunks']`, or specify `safe_chunks=False`."
+        )
+
+        for zchunk, dchunks, interval, size in zip(
+            enc_chunks_tuple, var_chunks, region, shape, strict=True
+        ):
+            if not safe_chunks:
+                continue
```
Comment (max-sixty) [inline xarray/backends/zarr.py:215]: nit: we could do an early return outside this loop

**Comment by max-sixty** [comment]
OK, very nice! (I haven't re-reviewed all the code, just the most recent commit).

There are still a couple of type errors; lmk if you need a hand with those.

**Comment by josephnowak** [comment]
> > Hi @max-sixty, sorry for being insistent with the review but I would like to avoid any misunderstood with the validation logic and affect the usability of Xarray.
> 
> (tbc, thank you very much for being insistent, this is a complicated and important problem which we haven't spent enough time on, so need exactly this sort of help!)
> 
> > The following case is going to raise an error using the mode "r+":
> 
> OK, that's a very good point — that writing to the full final chunk will raise an error with the proposed code. The rule should be that partial chunk writes are not allowed, but that is a full chunk write, so should be allowed. Otherwise people can't write to the full array!
> 
> Could we adjust the logic so writing to the final chunk is allowed?
> 
> > if it is, it probably would be good to make "a" the default mode, because it allows partial writes on the first and last chunk inside the region which I think is more convenient in more scenarios
> 
> I weigh the cost of unsafe writes very highly, so my initial impulse is that writing to partial chunks should fail by default.
> 
> Possibly making this strict could break some existing code; possibly we should explain ourselves better. For users who want to write to a single partial chunk from a single process, it would be a fairly small change for people to pass `mode="a"`, or allow unsafe writes.
> 
> But open to feedback from others; CC @pydata/xarray. One compromise would be to have a new mode, like `rd+`, for distributed writes, which would be strict like this. (though we should only do that if people feel strongly, otherwise we're at risk of config proliferation)

Yes, I can change the logic to avoid raising an error, I will have to send the shape of the Zarr array to validate if the region is covering the last chunk.

> > Hi @max-sixty, sorry for being insistent with the review but I would like to avoid any misunderstood with the validation logic and affect the usability of Xarray.
> 
> (tbc, thank you very much for being insistent, this is a complicated and important problem which we haven't spent enough time on, so need exactly this sort of help!)
> 
> > The following case is going to raise an error using the mode "r+":
> 
> OK, that's a very good point — that writing to the full final chunk will raise an error with the proposed code. The rule should be that partial chunk writes are not allowed, but that is a full chunk write, so should be allowed. Otherwise people can't write to the full array!
> 
> Could we adjust the logic so writing to the final chunk is allowed?
> 
> > if it is, it probably would be good to make "a" the default mode, because it allows partial writes on the first and last chunk inside the region which I think is more convenient in more scenarios
> 
> I weigh the cost of unsafe writes very highly, so my initial impulse is that writing to partial chunks should fail by default.
> 
> Possibly making this strict could break some existing code; possibly we should explain ourselves better. For users who want to write to a single partial chunk from a single process, it would be a fairly small change for people to pass `mode="a"`, or allow unsafe writes.
> 
> But open to feedback from others; CC @pydata/xarray. One compromise would be to have a new mode, like `rd+`, for distributed writes, which would be strict like this. (though we should only do that if people feel strongly, otherwise we're at risk of config proliferation)

I have already push 

> OK, very nice! (I haven't re-reviewed all the code, just the most recent commit).
> 
> There are still a couple of type errors; lmk if you need a hand with those.

I will check the type error, I thought that it was related to the code of someone else

**Comment by max-sixty** [comment]
Great! I'll merge, and if anyone has feedback we can make any adjustments later.

Thank you very much @josephnowak !

**Comment by shoyer** [comment]
Thanks Max and Joseph! These checks will save Xarray users a lot of grief

**Comment by shoyer** [comment]
Unfortunately this error seems to be triggered incorrectly in some cases: https://github.com/pydata/xarray/issues/9557

Please take a look if you have the time.

**Comment by shoyer** [comment]
We reverted this in https://github.com/pydata/xarray/pull/9558.

Would love to get this back in again after fixing https://github.com/pydata/xarray/issues/9557!

**Pull request merged**